modules/client-common/src/main/java/org/apache/ignite/client/proto/query/QueryEventHandlerImpl.java

```diff
@@ -0,0 +1,211 @@
+     */
+    public QueryEventHandlerImpl(QueryProcessor processor) {
+        this.processor = processor;
+    }
+
+    /** {@inheritDoc} */
+    @Override public JdbcQueryExecuteResult query(JdbcQueryExecuteRequest req) {
+        if (req.pageSize() <= 0)
+            return new JdbcQueryExecuteResult(JdbcResponse.STATUS_FAILED,
+                "Invalid fetch size : [fetchSize=" + req.pageSize() + ']');
+
+        var cursors = processor.query(req.schemaName(), req.sqlQuery(), req.arguments());
+
+        if (cursors.isEmpty())
+            return new JdbcQueryExecuteResult(JdbcResponse.STATUS_FAILED,
+                "At least one cursor is expected for query " + req.sqlQuery());
+
+        List<JdbcQuerySingleResult> results = new ArrayList<>();
+
+        try {
+            for (SqlCursor<List<?>> cur : cursors) {
+                JdbcQuerySingleResult res = createJdbcResult(cur, req);
+                results.add(res);
+            }
+        } catch (Exception ex) {
+            return new JdbcQueryExecuteResult(JdbcResponse.STATUS_FAILED,
+                "Failed to fetch results for query " + req.sqlQuery() + ". Error message: " + ex.getMessage());
+        }
+
+        return new JdbcQueryExecuteResult(results);
+    }
+
+    /** {@inheritDoc} */
+    @Override public JdbcQueryFetchResult fetch(JdbcQueryFetchRequest req) {
+        Cursor<List<?>> cur = state.get(req.cursorId());
+
+        if (cur == null)
+            return new JdbcQueryFetchResult(JdbcResponse.STATUS_FAILED,
+                "Failed to find query cursor with ID: " + req.cursorId());
+
+        if (req.pageSize() <= 0)
+            return new JdbcQueryFetchResult(JdbcResponse.STATUS_FAILED,
+                "Invalid fetch size : [fetchSize=" + req.pageSize() + ']');
+
+        List<List<Object>> fetch;
+        boolean hasNext;
+
+        try {
+            fetch = fetchNext(req.pageSize(), cur);
+            hasNext = cur.hasNext();
+        } catch (Exception ex) {
+            return new JdbcQueryFetchResult(JdbcResponse.STATUS_FAILED,
+                "Failed to fetch results for cursor id " + req.cursorId() + ". Error message: " + ex.getMessage());
+        }
+
+        return new JdbcQueryFetchResult(fetch, hasNext);
+    }
+
+    /** {@inheritDoc} */
+    @Override public JdbcBatchExecuteResult batch(JdbcBatchExecuteRequest req) {
+        return new JdbcBatchExecuteResult(UNSUPPORTED_OPERATION,
+            "ExecuteBatch operation is not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override public JdbcQueryCloseResult close(JdbcQueryCloseRequest req) {
+        Cursor<List<?>> cur = state.remove(req.cursorId());
+
+        if (cur == null)
+            return new JdbcQueryCloseResult(JdbcResponse.STATUS_FAILED,
+                "Failed to find query cursor with ID: " + req.cursorId());
+
+        try {
+            cur.close();
+        }
+        catch (Exception ex) {
+            state.remove(req.cursorId());
```

Comment:
what reason to remove an already removed cursor?

Reply:
Removed it, thanks.